forecasting/src/autogluon/forecasting/trainer/abstract_trainer.py

```diff
@@ -0,0 +1,668 @@
+        hyperparameters: Optional[Union[str, Dict]] = None,
+        models: Optional[List[AbstractForecastingModel]] = None,
+        val_data: Optional[TimeSeriesDataFrame] = None,
+        hyperparameter_tune: bool = False,
+        time_limit: Optional[float] = None,
+    ) -> List[str]:
+        time_start = time.time()
+        if hyperparameters is not None:
+            self.hyperparameters = hyperparameters
+            hyperparameters = copy.deepcopy(hyperparameters)
+        else:
+            if models is None:
+                raise ValueError("Either models or hyperparameters should be provided")
+
+        if self.save_data and not self.is_data_saved:
+            self.save_train_data(train_data)
+            if val_data is not None:
+                self.save_val_data(val_data)
+            self.is_data_saved = True
+
+        if models is None:
+            models = self.construct_model_templates(
+                hyperparameters, hyperparameter_tune=hyperparameter_tune
+            )
+
+        time_limit_model_split = time_limit
+        if time_limit is not None and len(models) > 0:
+            time_limit_model_split /= len(models)
+
+        model_names_trained = []
+        for i, model in enumerate(models):
+            if hyperparameter_tune:
+                time_left = time_limit_model_split
+            else:
+                if time_limit is None:
+                    time_left = None
+                else:
+                    time_start_model = time.time()
+                    time_left = time_limit - (time_start_model - time_start)
+
+            model_names_trained += self._train_single_full(
+                train_data,
+                model,
+                val_data=val_data,
+                hyperparameter_tune=hyperparameter_tune,
+                time_limit=time_left,
+            )
+
+        return model_names_trained
+
+    def leaderboard(self, data: Optional[TimeSeriesDataFrame] = None) -> pd.DataFrame:
+        logger.log(30, "Generating leaderboard for all models trained...")
+        model_names = self.get_model_names_all()
+        score_val = []
+        fit_time_marginal = []
+        fit_order = list(range(1, len(model_names) + 1))
+        score_dict = self.get_models_attribute_dict("val_score")
+        fit_time_marginal_dict = self.get_models_attribute_dict("fit_time")
+        for model_name in model_names:
+            score_val.append(score_dict[model_name])
+            fit_time_marginal.append(fit_time_marginal_dict[model_name])
+
+        test_score = []
+        if data is not None:
+            logger.log(
+                30, "Additional data provided, testing on the additional data..."
+            )
+            for model_name in model_names:
+                model: AbstractForecastingModel = self.load_model(model_name)
+                test_score.append(
+                    model.score(data) * METRIC_COEFFICIENTS[self.eval_metric]
+                )
```

Comment:
This kind of trick is going to stop working once we add ensembling, maybe instead move score logic to a separate method. such as `self.score(data, model)`, this way an ensemble model could preprocess the data to be able to properly calculate score via first calling predict on the base models.

Reply:
I revised to `self.score`, and added a todo in self score that reminds us this will break when ensembling is implemented.